### DISCUSSION TITLE
Using the provider for transactional operations

### DISCUSSION THREAD

**Discussion by bhwstevens3** [ueberdosis/hocuspocus#245] [Q&A]
When using the provider, Is there any method for performing one operation on the document and waiting for a confirmation from the server? 

I would like something like this:

```
async function updateMap(serverAddress, docName, key, value){
        // connect and do the update, resolve when we are SURE that it is done
        // throw error on problems
}

await updateMap('ws://127.0.0.1'. 'doc1', 'foo', 'bar');
```

After trying to do this via events, I could not find any event like "updated" or "update failed" in the current implementation...

**Comment by hanspagel** [comment]
Would you mind to describe why you’d like to use it like this?

**Reply by bhwstevens3** [reply]
@hanspagel We show rich widgets in our editor, a widget has some data in SQL, when I press the "duplicate" button, a request is sent to backend REST endpoint. It has to duplicate the database records first, and then duplicate parts of the Yjs document and finally let front know that it was successful with a HTTP 200 code.

This was just an example, currently there is no method for performing transactional operations without involving the provider. (like creating documents, updating documents or copying them)


The following issues are directly related to the root problem which is lack of access to the server without the provider:

https://github.com/ueberdosis/hocuspocus/discussions/142
https://github.com/ueberdosis/hocuspocus/issues/231
https://github.com/ueberdosis/hocuspocus/issues/222

I'd prefer not involving the provider at all and instead making some interface for direct access to the server functions (because currently, the provider is not designed for transactional operations and no other method is available).

**Reply by bhwstevens3** [reply]
BTW, I have currently bypassed this by adding the properties to an empty doc, and waiting for the "synced" event. But this will not work if someone wants to, for example, push something in an array! How would we know that it was successful?
